pkg/daemon/gateway_linux.go

```diff
@@ -415,6 +415,104 @@ func (c *Controller) deletePolicyRouting(family int, _ string, priority, tableID
 	return nil
 }
 
+// findRulePositions locates all occurrences of a rule in the specified table and chain,
+// searching from lowest to highest priority (bottom to top).
+// Returns all matching position indices, or an empty slice if none found.
+func findRulePositions(ipt *iptables.IPTables, rule util.IPTableRule) ([]int, error) {
+	rules, err := ipt.List(rule.Table, rule.Chain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list iptables rules: %w", err)
+	}
+
+	positions := make([]int, 0)
+
+	// Start from the end of the list (lowest priority) and go up to 1 (but skip index 0)
+	for i := len(rules) - 1; i >= 1; i-- {
+		ruleSpec := util.DoubleQuotedFields(rules[i])
+		if len(ruleSpec) < 3 {
+			continue
+		}
+
+		if slices.Equal(ruleSpec[2:], rule.Rule) {
+			positions = append(positions, i)
+		}
+	}
+
+	return positions, nil
+}
+
+// ensureNatPreroutingRulePosition ensures that the nat prerouting rule has a higher priority than the kube-proxy rule.
+func ensureNatPreroutingRulePosition(ipt *iptables.IPTables, rule util.IPTableRule) error {
+	kubeProxyRule := util.IPTableRule{
+		Table: "nat",
+		Chain: "PREROUTING",
+		Rule:  []string{"-m", "comment", "--comment", "kubernetes service portals", "-j", "KUBE-SERVICES"},
+	}
+	kubeProxyPosList, err := findRulePositions(ipt, kubeProxyRule)
+	if err != nil {
+		return fmt.Errorf("failed to find kube-proxy rule position: %w", err)
+	}
+
+	insertPosition := 1
+	if len(kubeProxyPosList) > 0 {
+		insertPosition = kubeProxyPosList[len(kubeProxyPosList)-1]
+	}
+
+	// Check if the rule already exists at a higher priority than the kube-proxy rule
+	existingNatPreroutingPosList, err := findRulePositions(ipt, rule)
+	if err != nil {
+		return fmt.Errorf("failed to find nat prerouting rule position: %w", err)
+	}
+	if len(existingNatPreroutingPosList) > 0 {
+		for _, pos := range existingNatPreroutingPosList {
+			if pos < insertPosition {
+				klog.V(5).Infof("nat prerouting rule already exists at position %d", pos)
+				return nil
+			}
+		}
+	}
+
+	klog.Infof("inserting nat prerouting rule %q at position %d", rule.Rule, insertPosition)
+	return ipt.Insert(rule.Table, rule.Chain, insertPosition, rule.Rule...)
+}
+
+// ensureNatPreroutingRuleNoDuplicate ensures that there are no duplicate nat prerouting rules.
+func ensureNatPreroutingRuleNoDuplicate(ipt *iptables.IPTables, rule util.IPTableRule) error {
+	existingNatPreroutingPosList, err := findRulePositions(ipt, rule)
+	if err != nil {
+		return fmt.Errorf("failed to find nat prerouting rule position: %w", err)
+	}
+	if len(existingNatPreroutingPosList) == 0 {
+		klog.Warningf("nat prerouting rule not found, skipping duplicate check")
+		return nil
+	}
+
```

Comment:
There is still a possibility that the iptables index will change between findRulePositions and the deletion/insertion of rules if Calico or other network operators insert rules at the same time. I looked through the go-iptables library but couldn't find any locking or transaction mechanism, so we may need to accept this limitation.

Reply:
Yes, it is unfortunately possible. I can add comments to state the limitation. FWIW, this would not bring damage when it comes to rule insertion, because the controller remove the duplicates later on.